pkg/registry/generic/registry/store.go

```diff
@@ -234,6 +235,36 @@ func (e *Store) Create(ctx api.Context, obj runtime.Object) (runtime.Object, err
 	return out, nil
 }
 
+// shouldDelete checks if a Update is removing all the object's finalizers. If so,
+// it further checks if the object's DeletionGracePeriodSeconds is 0. If so, it
+// returns true.
+func (e *Store) shouldDelete(ctx api.Context, key string, obj, existing runtime.Object) bool {
+	if err := e.Storage.Get(ctx, key, existing, false); err != nil {
+		utilruntime.HandleError(err)
+		return false
+	}
+	// Note: we do not check doUnconditionalUpdate in this routine, because we
+	// expect in most cases the update of finalizers will be made by non-human,
+	// and we expect the resourceVersion will be properly set.
+
+	// validate the update.
+	if err := rest.BeforeUpdate(e.UpdateStrategy, ctx, obj, existing); err != nil {
+		utilruntime.HandleError(err)
+		return false
+	}
+	newMeta, err := api.ObjectMetaFor(obj)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return false
+	}
+	oldMeta, err := api.ObjectMetaFor(existing)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return false
+	}
+	return len(oldMeta.Finalizers) != 0 && len(newMeta.Finalizers) == 0 && newMeta.DeletionGracePeriodSeconds != nil && *newMeta.DeletionGracePeriodSeconds == 0
```

Comment:
Hm.. idempotent is a good argument. I will remove that check.

Just for the record, my concern was if we remove the check of the old object having finalizers, then  we would allow people to delete object by updating the DeletionGracePeriods to 0, which would be a new behavior. But now I realized the concern was not valid, because the BeforeUpdate() on line 251 validated that the Update request doesn't change the DeletionGracePeriods.